domain/network/state/linklayer_merge.go

```diff
@@ -768,3 +818,155 @@ SET subnet_uuid = (
 		ProviderSubnetID: address.ProviderSubnetID,
 	}).Run()
 }
+
+// subnetCIDRUUIDByProviderID returns a map of subnet provider IDs to a
+// struct including the subnet CIDR and UUID.
+func (st *State) subnetCIDRUUIDByProviderID(
+	ctx context.Context,
+	tx *sqlair.TX,
+	add map[string][]mergeAddress,
+) (map[string]providerSubnetCIDR, error) {
+	type ids []string
+	input := make(ids, 0)
+	for _, addrs := range add {
+		for _, addr := range addrs {
+			input = append(input, addr.ProviderSubnetID)
+		}
+	}
+	stmt, err := st.Prepare(`
+SELECT &providerSubnetCIDR.*
+FROM provider_subnet AS ps
+JOIN subnet AS s ON ps.subnet_uuid = s.uuid
+WHERE ps.provider_id IN ($ids[:])
+`, providerSubnetCIDR{}, ids{})
+	if err != nil {
+		return nil, errors.Errorf("preparing subnet query: %w", err)
+	}
+
+	output := []providerSubnetCIDR{}
+	err = tx.Query(ctx, stmt, input).GetAll(&output)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, errors.Capture(err)
+	}
+
+	result := transform.SliceToMap(output, func(in providerSubnetCIDR) (string, providerSubnetCIDR) {
+		return in.ProviderID, in
+	})
+
+	return result, nil
+}
+
+// ensureAddressSubnetSuffix return a address including the subnet's suffix
+// and the subnetUUID. If the address is not an IPv4 nor IPv6 address, no
+// suffix is added.
+func ensureAddressSubnetSuffix(value, subnetProviderID string, data map[string]providerSubnetCIDR) (string, string, error) {
+	// Getting the subnet is helpful, but not essential.
+	subnet, _ := data[subnetProviderID]
+
+	parts := strings.Split(value, "/")
+	if len(parts) == 2 {
+		// Best case, the address has a CIDR suffix.
+		return subnet.SubnetUUID, value, nil
+	}
+
+	// Find the CIDR suffix, try the subnetCIDR first, if
+	// not define based on address type.
+	var suffix string
+	subnetCIDRParts := strings.Split(subnet.CIDR, "/")
+	switch len(subnetCIDRParts) {
+	case 2:
+		suffix = subnetCIDRParts[1]
+	case 1:
+		addType := corenetwork.DeriveAddressType(value)
+		if addType == corenetwork.IPv4Address {
+			suffix = "/32"
+		} else if addType == corenetwork.IPv6Address {
+			suffix = "/128"
+		} else {
+			return subnet.SubnetUUID, value, nil
+		}
+	case 0:
+		return "", "", errors.Errorf("invalid subnet CIDR value: %q", subnet.CIDR)
+	}
+
+	addressValue := value + "/" + suffix
+	return subnet.SubnetUUID, addressValue, nil
+}
+
+func (st *State) addAddressFromProvider(ctx context.Context, tx *sqlair.TX, add map[string][]mergeAddress) error {
```

Comment:
Which means this probably doesn't need to be a map.

Reply:
The `mergeAddress` doesn't have the device UUID in it. Passing the entire `mergeDevice` could include addresses which have already been created. Leaving as is.